src/phoenix/utilities/error_handling.py

```diff
@@ -0,0 +1,49 @@
+import logging
+from typing import Any, Callable, Iterable, Optional, Type
+
+from typing_extensions import TypeVar, cast
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+def graceful_fallback(
+    fallback_method: Callable[..., Any], exceptions: Optional[Iterable[Type[BaseException]]] = None
+) -> Callable[[F], F]:
+    """
+    Decorator that reroutes failing functions to a specified fallback method.
+
+    While it is generally not advisable to catch all exceptions, this decorator can be used to
+    gracefully degrade a function in situations when raising an error might be too disruptive.
+    Exceptions supprssed by this decorator will be logged to the root logger, and all inputs
+    to the wrapped function will be passed to the fallback method.
+
+
+    Args:
+
+    fallback_method (Callable[..., Any]): The fallback method to be called when the wrapped
+    function fails.
+
+    exceptions: An optional iterable of exceptions that should be suppressed by this decorator. If
+    unset, all exceptions will be suppressed.
+    """
+
+    exceptions = (BaseException,) if exceptions is None else tuple(exceptions)
+
+    def decorator(func: F) -> F:
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            try:
+                return func(*args, **kwargs)
+            except exceptions as exc:
+                msg = (
+                    f"Exception occurred in function '{func.__name__}':\n"
+                    f"  Args: {args}\n"
+                    f"  Kwargs: {kwargs}\n"
+                    f"  Exception Type: {type(exc).__name__}\n"
+                    f"  Exception Message: {str(exc)}"
+                )
+                logging.error(msg)
```

Comment:
It seems like the fallback method is currently handling two responsibilities, logging error information and returning a fallback value, but logging the error is split in two places. Can we consolidate the error logging into a single place?

For example, instead of logging the exception information here, what if we delegate that responsibility to the `fallback_method`, which might be called an `error_handler`? That way, there would only be a single error message logged and all the error logging code is in the same place.

Reply:
I think it's important that the fallback decorator handles both: since falling back to a new method only happens when an error is suppressed. I think it's better for logging to happen alongside error suppression as opposed to leaving it to the fallback method.

I agree that the double logging might be a little bit ugly, and I'm thinking about three ways we might be able to resolve it:

1. remove the logging in the fallback method: I don't think this one is necessary anyway and I wanted to preserve the message that was already in our codebase. I don't think we could ever *prevent* someone from wanting to log additional information so this is kind of a specific solution. If we really don't want two logs this is what I like best.
2. add a new argument to the decorator where we can shim in additional logging information: I don't like this so much, as you mentioned the logger is responsible for a few concepts
3. rely on the fallback method to log entirely: I don't like this, as noted above ignoring an error means we should know what to do with that error (log it) and how to proceed (falling back)

As written though, I don't think two logs are so bad, so I didn't pick any of these. What do you think?